contracts/vesting/src/contract.rs

```diff
@@ -0,0 +1,526 @@
+    //         ),
+    //         (owner, contract, amount),
+    //     );
+
+    //     Ok(())
+    // }
+
+    fn update_minter(env: Env, sender: Address, new_minter: Address) {
+        let current_minter = get_minter(&env);
+
+        let is_authorized = if let Some(current_minter) = current_minter.clone() {
+            sender == current_minter.address
+        } else {
+            sender == get_admin(&env)
+        };
+
+        if !is_authorized {
+            log!(
+                env,
+                "Vesting: Update minter: Not authorized to update minter"
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        let mint_capacity = current_minter.map_or(0, |m| m.mint_capacity);
+        save_minter(
+            &env,
+            &MinterInfo {
+                address: new_minter.clone(),
+                mint_capacity,
+            },
+        );
+
+        env.events()
+            .publish(("Update minter", "Updated minter to: "), new_minter);
+    }
+
+    fn update_minter_capacity(env: Env, sender: Address, new_capacity: u128) {
+        if sender != get_admin(&env) {
+            log!(
+                &env,
+                "Vesting: Update minter capacity: Only contract's admin can update the minter's capacity"
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        if let Some(minter) = get_minter(&env) {
+            save_minter(
+                &env,
+                &MinterInfo {
+                    address: minter.address,
+                    mint_capacity: new_capacity,
+                },
+            );
+        } else {
+            log!(&env, "Vesting: Update Minter Capacity: Minter not found");
+            panic_with_error!(env, ContractError::MinterNotFound);
+        };
+
+        env.events().publish(
+            ("Update minter capacity", "Updated minter capacity to: "),
+            new_capacity,
+        );
+    }
+
+    fn query_balance(env: Env, address: Address) -> i128 {
+        token_contract::Client::new(&env, &get_token_info(&env).address).balance(&address)
+    }
+
+    fn query_distribution_info(
+        env: Env,
+        address: Address,
+    ) -> Result<DistributionInfo, ContractError> {
+        Ok(get_vesting(&env, &address)?.distribution_info)
+    }
+
+    fn query_token_info(env: Env) -> VestingTokenInfo {
+        get_token_info(&env)
+    }
+
+    fn query_minter(env: Env) -> MinterInfo {
+        if let Some(minter) = get_minter(&env) {
+            minter
+        } else {
+            log!(&env, "Vesting: Query Minter: Minter not found");
+            panic_with_error!(env, ContractError::MinterNotFound);
+        }
+    }
+
+    fn query_vesting_total_supply(env: Env) -> i128 {
+        let token_address = get_token_info(&env).address;
+        token_contract::Client::new(&env, &token_address).balance(&env.current_contract_address())
+    }
```

Comment:
What is missing in those queries is to have a clear point where user can see how many tokens he can withdraw.

Reply:
added `query_available_to_claim` and `claim` so users can automatically claim whatever is available for them